src/main/java/pl/javastart/restoffers/offer/OfferService.java

```diff
@@ -0,0 +1,114 @@
+package pl.javastart.restoffers.offer;
+
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.stereotype.Service;
+import pl.javastart.restoffers.category.CategoryRepository;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+public class OfferService {
+    private final OfferRepository offerRepository;
+    private final CategoryRepository categoryRepository;
+
+    public OfferService(OfferRepository offerRepository, CategoryRepository categoryRepository) {
+        this.offerRepository = offerRepository;
+        this.categoryRepository = categoryRepository;
+    }
+
+    public List<OfferDto> findFiltered(Long categoryId, String title) {
+        List<Offer> offers;
+        if (categoryId == null) {
+            if (title == null) {
+                offers = offerRepository.findAll();
+            } else {
+                offers = offerRepository.findAll()
+                        .stream()
+                        .filter(offer -> offer.getTitle().toLowerCase()
+                                .contains(title.toLowerCase()))
+                        .collect(Collectors.toList());
+            }
+        } else {
+            if (title == null) {
+                offers = offerRepository.findByCategoryId(categoryId);
+            } else {
+                offers = offerRepository.findByCategoryId(categoryId)
+                        .stream()
+                        .filter(offer -> offer.getTitle().toLowerCase()
+                                .contains(title.toLowerCase()))
+                        .collect(Collectors.toList());
+            }
+        }
```

Comment:
Filtrowanie po tytule się tutaj powtarza, można to trochę inaczej zapisać, pokazane w code review